src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java

```diff
@@ -0,0 +1,240 @@
+ * <a href="https://checkstyle.org/config.html#tabWidth">tabWidth</a> property.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code violateExecutionOnNonTightHtml} - Control when to print violations if the
+ * Javadoc being examined by this check violates the tight html rules defined at
+ * <a href="https://checkstyle.org/writingjavadocchecks.html#Tight-HTML_rules">Tight-HTML Rules</a>.
+ * Type is {@code boolean}.
+ * Default value is {@code false}.
+ * </li>
+ * </ul>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code javadoc.asterisk.indentation}
+ * </li>
+ * <li>
+ * {@code javadoc.missed.html.close}
+ * </li>
+ * <li>
+ * {@code javadoc.parse.rule.error}
+ * </li>
+ * <li>
+ * {@code javadoc.unclosedHtml}
+ * </li>
+ * <li>
+ * {@code javadoc.wrong.singleton.html.tag}
+ * </li>
+ * </ul>
+ *
+ * @since 10.17.0
+ */
+@GlobalStatefulCheck
+public class JavadocLeadingAsteriskAlignCheck extends AbstractJavadocCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "javadoc.asterisk.indentation";
+
+    /** Specifies the column number of starting block of the javadoc comment with tabs expanded. */
+    private int expectedColumnNumber;
+
+    /** Specifies the line number of starting block of the javadoc comment. */
+    private int javadocStartLineNumber;
+
+    /**
+     * Specifies the column number of the leading asterisk
+     * where tabs are converted to a single whitespace.
+     */
+    private int singleSpaceAsteriskColNumber;
+
+    /** Specifies the lines of the file being processed. */
+    private String[] fileLines;
+
+    @Override
+    public int[] getDefaultJavadocTokens() {
+        return new int[] {
+            JavadocTokenTypes.LEADING_ASTERISK,
+        };
+    }
+
+    @Override
+    public int[] getRequiredJavadocTokens() {
+        return getAcceptableJavadocTokens();
+    }
+
+    @Override
+    public void beginJavadocTree(DetailNode rootAst) {
+        // this method processes and sets information of starting javadoc tag.
+        fileLines = getLines();
+        final String startLine = fileLines[rootAst.getLineNumber() - 1];
+        javadocStartLineNumber = rootAst.getLineNumber();
+        expectedColumnNumber = CommonUtil.lengthExpandedTabs(
+            startLine, rootAst.getColumnNumber() - 1, getTabWidth());
+    }
+
+    @Override
+    public void visitJavadocToken(DetailNode ast) {
+        // this method checks the alignment of leading asterisks.
+        final boolean isJavadocStartingLine = ast.getLineNumber() == javadocStartLineNumber;
+        if (!isJavadocStartingLine) {
+            final OptionalInt asteriskColumnNumber =
+                                        getAsteriskColumnNumberWithTabsExpanded(ast.getText());
+
+            asteriskColumnNumber.ifPresent(integer -> {
```

Comment:
Here we should map to expand tabs, then filter on `hasValidAlignment`. Like this (this is pseudocode):
```suggestion
            leadingAsteriskColumnNumber
                .map(columnNumber -> expandTabs(line, columnNumber))
                .filter(expanded -> hasValidAlignment(expected, expanded))
                .ifPresent(violation -> logViolation(violation, expected));
```

Reply:
The goal here is to make what we are doing clear, but not reveal _how_ we are doing it; additionally, we want to avoid calling methods unnecessarily. Preventing extra method calls helps us to avoid adding weird conditions in the future, as well as laying down good design to iterate upon as we add new check features. We want to expose enough here to be flexible going forward, but also keep the entry point to this check easy to understand.

Reply:
Done. here is the new changes:

https://github.com/checkstyle/checkstyle/blob/77dd1d104385ad79133495efee9f504f23f06730/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L139-L152

https://github.com/checkstyle/checkstyle/blob/77dd1d104385ad79133495efee9f504f23f06730/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L163-L176